lib/core/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs

```diff
@@ -447,6 +451,57 @@ spec = do
                             . slotStartTime sps
                     Just slot === f slot
 
+        it "slotRangeFromTimeRange is maximal" $
+            property $ \(sps, uniformTimeRange) ->
+
+                let timeRange :: Range UTCTime
+                    timeRange = getUniformTime <$> uniformTimeRange
+
+                    maybeSlotRange :: Maybe (Range SlotId)
+                    maybeSlotRange = slotRangeFromTimeRange sps timeRange
+
+                    startsWithin :: Range SlotId -> Range UTCTime -> Bool
+                    startsWithin sr tr =
+                        (`isSubrangeOf` tr) $ fmap (slotStartTime sps) sr
+
+                    lowerBoundPred = mapRangeLowerBound slotPred'
+                    upperBoundSucc = mapRangeUpperBound slotSucc'
+
+                    slotPred' :: SlotId -> SlotId
+                    slotPred' s = fromMaybe slotMinBound $ slotPred sps s
+
+                    slotSucc' :: SlotId -> SlotId
+                    slotSucc' = slotSucc sps
+                in
+                checkCoverage $
+                cover 20 (isNothing maybeSlotRange)
+                    "have no slot range" $
+                cover 50 (isJust maybeSlotRange)
+                    "have slot range" $
+                cover 20 (fmap rangeHasLowerBound maybeSlotRange == Just True)
+                    "slot range has lower bound" $
+                cover 20 (fmap rangeHasUpperBound maybeSlotRange == Just True)
+                    "slot range has upper bound" $
+                cover 20 (fmap rangeIsFinite maybeSlotRange == Just True)
+                    "slot range is finite" $
+
+                case maybeSlotRange of
+                    Nothing ->
+                        (Just True ===
+                            ((< getGenesisBlockDate sps) . StartTime
+                                <$> inclusiveUpperBound timeRange))
+                    Just slotRange ->
+                        (slotRange `startsWithin` timeRange
+                            === True)
+                        .&&.
+                        (lowerBoundPred slotRange `startsWithin` timeRange
+                            === (not (rangeHasLowerBound slotRange)
+                                || lowerBoundPred slotRange == slotRange))
```

Comment:
I can't quite follow here. Wouldn't  `(not (rangeHasLowerBound slotRange)` the same as `lowerBoundPred slotRange == slotRange`?

Reply:
> I can't quite follow here. Wouldn't `(not (rangeHasLowerBound slotRange)` the same as `lowerBoundPred slotRange == slotRange`?

Not quite!

The intention of this test is that ```lowerBoundPred slotRange `startsWithin` timeRange``` should generally be `False`, but we allow it to be `True` if either (or both) of two conditions are satisfied:

1. **The range doesn't have a lower bound**
    (in which case, `lowerBoundPred` will be the same as `id`).
2. **The range has a lower bound of `SlotId 0 0`**
    (in which case `lowerBoundPred` will be the same as `id`).

Let's consider the following example:
```hs
slotRange = Range (Just (SlotId 0 0)) Nothing
```

For the first condition, we get the following result (by substitution):
```hs
  not (rangeHasLowerBound slotRange)
= not (rangeHasLowerBound (Range (Just (SlotId 0 0)) Nothing))
= not (isJust (Just (SlotId 0 0)))
= not True
= False
```

For the second condition we get:
```hs
  slotRange == lowerBoundPred slotRange
= slotRange == lowerBoundPred (Range (Just (SlotId 0 0)) Nothing)
= slotRange == Range (Just (slotPred' (SlotId 0 0))) Nothing
= slotRange == Range (Just (fromMaybe slotMinBound (slotPred (SlotId 0 0)))) Nothing
= slotRange == Range (Just (fromMaybe slotMinBound Nothing)) Nothing
= slotRange == Range (Just slotMinBound) Nothing
= slotRange == Range (Just (SlotId 0 0)) Nothing
= slotRange == slotRange
= True
```

What you might have missed is that we're using `slotPred'` (which has type `SlotId -> SlotId`) rather than `slotPred` (which has type `SlotId -> Maybe SlotId`).